src/runtime/list.cpp

```diff
@@ -63,36 +61,41 @@ extern "C" PyObject* PyList_AsTuple(PyObject* v) noexcept {
 }
 
 extern "C" Box* listRepr(BoxedList* self) {
-    // TODO highly inefficient with all the string copying
-
-    std::string O("");
-    llvm::raw_string_ostream os(O);
-
-    // Implementing Recursive Print of list in same way from  Cpython
-    int recursive = Py_ReprEnter((PyObject*)self);
-    if (recursive != 0) {
-        if (recursive < 0)
-            return boxString(os.str());
+    std::vector<char> chars;
+    int status = Py_ReprEnter((PyObject*)self);
 
-        os << "[...]";
-        return boxString(os.str());
+    if (status != 0) {
+        try {
+            if (status < 0)
+                throwCAPIException();
+        } catch (ExcInfo e) {
+            Py_ReprLeave((PyObject*)self);
+            throw e;
+        }
+        chars.push_back('[');
+        chars.push_back('.');
+        chars.push_back('.');
+        chars.push_back('.');
+        chars.push_back(']');
+        return boxString(llvm::StringRef(&chars[0], chars.size()));
     }
+    chars.push_back('[');
 
-    os << '[';
     for (int i = 0; i < self->size; i++) {
-        if (i > 0)
-            os << ", ";
 
+        if (i > 0) {
+            chars.push_back(',');
+            chars.push_back(' ');
+        }
         Box* r = self->elts->elts[i]->reprICAsString();
```

Comment:
This can throw so we need to put it in a try-catch block -- I'd start the try right after the end of the `if (status != 0)` block, and put the end (ie the catch) right before the normal call to Py_ReprLeave.